src/storage/segcache/src/ttl_buckets/mod.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use crate::common::ThinOption;
+use crate::*;
+
+use rustcommon_time::CoarseInstant as Instant;
+
+#[cfg(feature = "dump")]
+use serde::{Deserialize, Serialize};
+
+mod constants;
+mod error;
+
+pub use error::Error;
+
+use constants::*;
+
+pub struct TtlBucket {
+    head: i32,
+    tail: i32,
+    ttl: i32,
+    segments: i32,
+    next_to_merge: i32,
+    _pad: [u8; 44],
+}
+
+pub struct TtlBuckets {
+    pub(crate) buckets: Box<[TtlBucket]>,
+}
+
+#[cfg(feature = "dump")]
+#[derive(Serialize, Deserialize)]
+pub struct TtlBucketDump {
+    ttl: i32,
+    head: i32,
+}
+
+impl TtlBucket {
+    fn new(ttl: i32) -> Self {
+        Self {
+            head: -1,
```

Comment:
We actually have enough bytes in the TtlBucket to use proper Option types - but they feel a little weird to use with a `i32` that treats negative values as a None variant. I wonder if using Option<core::num::NonZeroU32> (which fits in 32bits) would be more ergonomic. Does mean we'd be starting any index at 1, so might make offset calculating a little more awkward - but it would do away with some of the more bespoke choices like my 'ThinOption<i32>' type

Reply:
I think that's a fair tradeoff. And maybe that's a good practice anyway since we can also treat segment 0 as the "boot" zone or corruption-checking buffer (e.g. put magic values in there)? On the other hand, best practices on pointer-arithmetic or equivalent means containing them as much as possible anyway.

Reply:
Pushed the type change for segment ids - I think it works pretty well. I left the aggregate allocation the same - so the slice runs 0-(N-1) and the logical segment ids are 1-N.